Increase rcl_yaml_param_parser test coverage

brawner opened the pull request:
This adds more API specific tests for the rcl_yaml_yaram_parser. It makes use of a failing allocator that I'm calling the time bomb allocator because I required something that could succeed for a few calls and then fail. 

I do have some concerns that such a setup might be "too clever for its own good". However, it did help me find several initialization/finalization order bugs. [this guy](https://github.com/ros2/rcl/compare/brawner/api-test-coverage?expand=1#diff-94d86be26e23f7eb14a760089ff1fd26R547), would have been impossible to find without it.

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=10816)](http://ci.ros2.org/job/ci_linux/10816/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=6194)](http://ci.ros2.org/job/ci_linux-aarch64/6194/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=8807)](http://ci.ros2.org/job/ci_osx/8807/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=10703)](http://ci.ros2.org/job/ci_windows/10703/)

brawner commented:
Addressed uncrustify error:
[![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_linux&build=10817)](https://ci.ros2.org/job/ci_linux/10817/)

ahcorde commented on rcl_yaml_param_parser/test/test_parser.cpp:
```diff
@@ -0,0 +1,351 @@
+  EXPECT_EQ(nullptr, rcl_yaml_node_struct_copy(params_st));
+
+  for (int i = 0; i < 5; ++i) {
+    set_time_bomb_allocator_calloc_count(params_st->allocator, 4);
```
Why are you doing this 5 times? or:

```suggestion
    set_time_bomb_allocator_calloc_count(params_st->allocator, i);
```

brawner replied:
Fixed.

ahcorde commented on rcl_yaml_param_parser/test/test_parser.cpp:
```diff
@@ -0,0 +1,351 @@
+  set_time_bomb_allocator_malloc_count(params_st->allocator, num_malloc_calls_until_copy_param);
+  params_st->params->parameter_values[0].bool_value = nullptr;
+  params_st->params->parameter_values[0].integer_value = &temp_int;
+  EXPECT_EQ(nullptr, rcl_yaml_node_struct_copy(params_st));
```
if the time bomb allocator is set to `num_malloc_calls_until_copy_param`, why is this returning `nullptr`? I would expect to get the parameter copy

brawner replied:
Because that's how many malloc calls occur before it attempts to copy, which requires a malloc. So it should fail after that many calls.

ahcorde commented on rcl_yaml_param_parser/test/test_parser.cpp:
```diff
@@ -0,0 +1,351 @@
+      params_hdl = NULL;
+    }
+  }
+  SUCCEED();
```
is this need it?

brawner replied:
Removed

brawner commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=10851)](http://ci.ros2.org/job/ci_linux/10851/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=6221)](http://ci.ros2.org/job/ci_linux-aarch64/6221/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=8839)](http://ci.ros2.org/job/ci_osx/8839/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=10745)](http://ci.ros2.org/job/ci_windows/10745/)

brawner commented on rcl_yaml_param_parser/src/parser.c:
```diff
@@ -364,28 +365,38 @@ rcl_params_t * rcl_yaml_node_struct_init(
 
+  for (size_t i = 0; i < MAX_NUM_NODE_ENTRIES; ++i) {
+    params_st->node_names[i] = NULL;
+  }
```
You're right. As I was trying to nail down memory errors, I was trying to attempting to solve this with brute force. This was not needed and has been removed.

brawner commented on rcl_yaml_param_parser/src/parser.c:
```diff
@@ -364,28 +365,38 @@ rcl_params_t * rcl_yaml_node_struct_init(
+    params_st->params[i].parameter_names = NULL;
+    params_st->params[i].parameter_values = NULL;
+    params_st->params[i].num_params = 0;
+  }
```
Yup. This was removed.

brawner commented on rcl_yaml_param_parser/src/parser.c:
```diff
@@ -267,7 +267,7 @@ static rcutils_ret_t rem_name_from_ns(
       next_str = strstr(cur_ns, sep_str);
```
Split into two commits, one that is a cleanup and bug fixes, and the second that adds the test.

brawner commented on rcl_yaml_param_parser/test/test_parser.cpp:
```diff
@@ -0,0 +1,351 @@
+  const char node_name[] = "node name";
+  const char param_name[] = "param name";
+  const char yaml_value[] = "true";
+  EXPECT_TRUE(rcl_parse_yaml_value(&node_name[0], &param_name[0], &yaml_value[0], params_st));
```
Fixed everywhere. Don't know why I thought I needed those.

brawner commented on rcl_yaml_param_parser/test/test_parser.cpp:
```diff
@@ -0,0 +1,351 @@
+  // Check integer value
+  int64_t temp_int = 42;
+  set_time_bomb_allocator_malloc_count(params_st->allocator, num_malloc_calls_until_copy_param);
+  params_st->params->parameter_values[0].bool_value = nullptr;
```
You have the correct impression. Deallocating only the bool_value, since the others are allocated on the stack.

brawner commented on rcl_yaml_param_parser/test/test_parser.cpp:
```diff
@@ -0,0 +1,351 @@
+  // Check double value
+  double temp_double = 42.0;
+  set_time_bomb_allocator_malloc_count(params_st->allocator, num_malloc_calls_until_copy_param);
+  params_st->params->parameter_values[0].integer_value = nullptr;
```
Moved these lines a bit closer to where they are set.

brawner commented on rcl_yaml_param_parser/test/test_parser.cpp:
```diff
@@ -0,0 +1,351 @@
+  char temp_string[] = "stringy string";
+  set_time_bomb_allocator_malloc_count(params_st->allocator, num_malloc_calls_until_copy_param);
+  params_st->params->parameter_values[0].double_value = nullptr;
+  params_st->params->parameter_values[0].string_value = &temp_string[0];
```
Fixed.

brawner commented on rcl_yaml_param_parser/test/test_parser.cpp:
```diff
@@ -0,0 +1,351 @@
+  const char node_name[] = "node name";
+  const char param_name[] = "param name";
+  const char yaml_value[] = "true";
+  const char empty_string[1] = {'\0'};
```
Fixed.

brawner commented on rcl_yaml_param_parser/test/test_parser.cpp:
```diff
@@ -0,0 +1,351 @@
+  EXPECT_EQ(nullptr, rcl_yaml_node_struct_get(&node_name[0], nullptr, params_st));
+  EXPECT_EQ(nullptr, rcl_yaml_node_struct_get(&node_name[0], &param_name[0], nullptr));
+
+  rcl_variant_t * result = rcl_yaml_node_struct_get(&node_name[0], &param_name[0], params_st);
```
Fixed.

brawner commented on rcl_yaml_param_parser/test/test_parser.cpp:
```diff
@@ -0,0 +1,351 @@
+    bool res = rcl_parse_yaml_file(path, params_hdl);
+    // Not checking res here, because eventually it will come back with an ok result.
+    // We're just trying to make sure that bad allocations are properly handled
+    EXPECT_TRUE(res || !res);
```
I needed to use the result so I wouldn't get a warning. The if statement makes this EXPECT_TRUE pointless and it has since been removed.

brawner commented on rcl_yaml_param_parser/test/test_parser.cpp:
```diff
@@ -0,0 +1,351 @@
+    bool res = rcl_parse_yaml_file(path, params_hdl);
+    // Not checking res here, because eventually it will come back with an ok result.
+    // We're just trying to make sure that bad allocations are properly handled
+    EXPECT_TRUE(res || !res);
```
Fixed.

brawner commented on rcl_yaml_param_parser/test/test_parser.cpp:
```diff
@@ -0,0 +1,351 @@
+  EXPECT_EQ(nullptr, rcl_yaml_node_struct_get(&node_name[0], &param_name[0], nullptr));
+
+  rcl_variant_t * result = rcl_yaml_node_struct_get(&node_name[0], &param_name[0], params_st);
+  EXPECT_NE(nullptr, result->bool_value);
```
Fixed.

brawner commented:
I added in another for-loop with the time-bomb allocator, but this time for calloc calls. Since just about everything uses zero_allocate, it exposed a lot more issues. I traced down every memory error using valgrind, so now the tests run memory leak free!

I've also split the PR into two commits, one for the parser.c changes and the second for the test file changes. Unfortunately, that means that the history was deleted.

We're now up to 88% coverage! https://ci.ros2.org/job/ci_linux/10851/cobertura/src_ros2_rcl_rcl_yaml_param_parser_src/

brawner commented on rcl_yaml_param_parser/test/test_parse_yaml.cpp:
```diff
@@ -74,6 +78,8 @@ TEST(test_parser, correct_syntax) {
     ASSERT_TRUE(NULL != param_value->bool_value);
     EXPECT_TRUE(*param_value->bool_value);
+    allocator.deallocate(param_value->bool_value, allocator.state);
+    param_value->bool_value = nullptr;
```
These were removed, and cleanup code was added to parser.c. Back to 0 memory leaks!

hidmic commented on rcl_yaml_param_parser/src/parser.c:
```diff
@@ -1422,6 +1509,11 @@ static rcutils_ret_t parse_key(
           memmove((param_name + params_ns_len + 1U), value, param_name_len);
           param_name[tot_len - 1U] = '\0';
 
+          if (params_st->params[*node_idx].parameter_names[*parameter_idx] != NULL) {
```
@brawner nit^2: invert operands order per style guide, as in
```suggestion
          if (NULL != params_st->params[*node_idx].parameter_names[*parameter_idx]) {
```

brawner replied:
Fixed.

hidmic commented on rcl_yaml_param_parser/test/test_parser.cpp:
```diff
@@ -0,0 +1,363 @@
+  // Check integer value
+  int64_t temp_int = 42;
+  if (params_st->params->parameter_values[0].bool_value != nullptr) {
+    // Since this bool was allocated above in rcl_parse_yaml_value, it needs to be freed.
```
I see. Disregard.

hidmic commented:
On a related note, I've already seen failing allocators proliferating in many packages. The same is going to happen with the time bomb allocator. 

I think we should find a way to reuse these utilities (instead of replicating N times).

brawner commented:
Running CI one more time:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=10887)](http://ci.ros2.org/job/ci_linux/10887/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=6245)](http://ci.ros2.org/job/ci_linux-aarch64/6245/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=8867)](http://ci.ros2.org/job/ci_osx/8867/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=10778)](http://ci.ros2.org/job/ci_windows/10778/)

brawner commented:
I was thinking something along the same lines, maybe there can be an rcl_testing_utils package or whatnot. I'm glad to see you think my crazy time-bomb allocator might actually be useful to others!

hidmic commented:
> I'm glad to see you think my crazy time-bomb allocator might actually be useful to others!

I already need it to bump `rmw_dds_common` coverage past 91% :sweat_smile:
